File: DoubleB/ViewControllers/DBCompaniesViewController.m
//
//  DBCompaniesViewController.m
//  
//
//  Created by Balaban Alexander on 24/08/15.
//
//

#import "DBCompaniesViewController.h"
#import "AppDelegate.h"

#import "DBServerAPI.h"
#import "DBCompanyInfo.h"
#import "MBProgressHUD.h"
#import "DBCompaniesManager.h"

@interface DBCompaniesViewController () <UITableViewDelegate, UITableViewDataSource>

@property (strong, nonatomic) IBOutlet UITableView *tableView;

@property (strong, nonatomic) IBOutlet UIView *titleView;
@property (strong, nonatomic) IBOutlet UILabel *titleLabel;
@property (strong, nonatomic) IBOutlet NSLayoutConstraint *titleViewHeightConstraint;

@property (strong, nonatomic) NSArray *companies;

@end

@implementation DBCompaniesViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setNeedsStatusBarAppearanceUpdate];
    [self initializeViews];
    
    self.companies = [[DBCompaniesManager sharedInstance] companies];
}

- (void)initializeViews {
    if (self.mode == DBCompaniesViewControllerModeChangeCompany) {
        self.titleView.hidden = false;
        self.titleView.backgroundColor = [UIColor db_defaultColor];
    } else {
        self.titleView.hidden = true;
        self.titleViewHeightConstraint.constant = 0;
        [self.view layoutIfNeeded];
    }
    
    self.tableView.tableFooterView = [[UIView alloc] initWithFrame:CGRectZero];
    self.titleLabel.text = NSLocalizedString(@"Выберите ресторан", nil);
    self.titleLabel.textColor = [UIColor whiteColor];
    
    self.title = NSLocalizedString(@"Выберите ресторан", nil);
    [self.navigationController.navigationBar setTitleTextAttributes:@{NSForegroundColorAttributeName: [UIColor whiteColor]}];
}

- (void)putSelectedNamespace:(NSString *)namespace {
    [[ApplicationManager sharedInstance] flushStoredCache];
    [DBCompaniesManager selectCompanyName:namespace];
    
    [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    [[DBCompanyInfo sharedInstance] updateInfo:^(BOOL success) {
        [MBProgressHUD hideAllHUDsForView:self.view animated:YES];
        
        if(success){
            AppDelegate *delegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
            delegate.window.rootViewController = [ViewControllerManager mainViewController];
        } else {
            [self showError:@"Не удалось загрузить информацию о выбранное компании"];
        }
    }];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.companies.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [UITableViewCell new];
    cell.textLabel.text = [self.companies[indexPath.row] objectForKey:@"name"];
    return cell;
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [self putSelectedNamespace:[self.companies[indexPath.row] objectForKey:@"namespace"]];
}

@end

File: DoubleB/ViewControllers/DBCompaniesViewController.swift
//
//  DBCompaniesViewController.swift
//  
//
//  Created by Balaban Alexander on 18/06/15.
//
//

import UIKit

@objc
public class DBCompaniesViewController: UIViewController {
    @IBOutlet var tableView: UITableView!
    
    @IBOutlet var titleView: UIView!
    @IBOutlet var titleLabel: UILabel!
    @IBOutlet var titleViewHeightConstraint: NSLayoutConstraint!
    @IBOutlet weak var splashImageView: UIImageView!
    
    public var firstLaunch = true
    private var companies = [NSDictionary]()

    override public func viewDidLoad() {
        super.viewDidLoad()

        self.setNeedsStatusBarAppearanceUpdate()
        initializeViews()
        requestCompanies()
    }
    
    func initializeViews() {
        if firstLaunch {
            self.titleView.hidden = false
            self.titleView.backgroundColor = UIColor.db_defaultColor()
        } else {
            self.titleView.hidden = true
            titleViewHeightConstraint.constant = 0
            self.view.layoutIfNeeded()
        }
        self.tableView.tableFooterView = UIView(frame: CGRectZero)
        self.titleLabel.text = NSLocalizedString("Выберите ресторан", comment: "")
        self.titleLabel.textColor = UIColor.whiteColor()
        self.title = NSLocalizedString("Выберите ресторан", comment: "")
        self.navigationController?.navigationBar.titleTextAttributes = [NSForegroundColorAttributeName: UIColor.whiteColor()]
        
        let height = Int(UIScreen.mainScreen().bounds.size.height)
        let launchScreenName = "launch_\(height).png"
        if self.firstLaunch {
            self.splashImageView.image = UIImage(named:launchScreenName)
        }
    }
    
    func requestCompanies() {
        DBServerAPI.requestCompanies({ (companies) -> Void in
            self.companies = companies as! [NSArray]
            self.tableView.reloadData()
            if self.companies.count == 1 {
                self.putSelectedNamespace(self.companies.first!.objectForKey("namespace") as! String)
            } else {
                self.splashImageView.hidden = true
            }
        }, failure: { (error) -> Void in
            
        })
    }
    
    func putSelectedNamespace(namespace: String) {
        DBCompaniesManager.sharedInstance().selectedCompanyName = namespace
        
        let delegate = UIApplication.sharedApplication().delegate as! AppDelegate
        self.preloadData()
        if (DBCompanyInfo.sharedInstance().deliveryTypes != nil) {
            delegate.window.rootViewController = LaunchViewController()
        } else {
            delegate.window.rootViewController = DBTabBarController.sharedInstance()
        }
    }
    
    public override func prefersStatusBarHidden() -> Bool {
        return firstLaunch
    }
}

extension DBCompaniesViewController: UITableViewDataSource {
    public func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.companies.count
    }
    
    public func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        var cell = UITableViewCell()
        cell.textLabel!.text = self.companies[indexPath.row].objectForKey("name") as? String
        return cell
    }
}

extension DBCompaniesViewController: UITableViewDelegate {
    public func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let namespace = self.companies[indexPath.row].objectForKey("namespace") as! String
        putSelectedNamespace(namespace)
    }
    
    func preloadData() {
        DBServerAPI.registerUser(nil)
        Venue.dropAllVenues()
        Venue.fetchAllVenuesWithCompletionHandler { (venues) -> Void in
            let delegate = UIApplication.sharedApplication().delegate as! AppDelegate
            delegate.saveContext()
        }
        OrderManager.sharedManager().reset()
        DBDeliverySettings.sharedInstance().reset()
        DBMenu.sharedInstance().clearMenu()
        DBMenu.sharedInstance().updateMenuForVenue(nil, remoteMenu: nil)
        Order.dropAllOrders()
        DBPromoManager.sharedManager().clear()
        DBPromoManager.sharedManager().updateInfo()
        DBTabBarController.sharedInstance().moveToStartState()
        DBCompanyInfo.sharedInstance().updateInfo()
    }
}